SYCL/Reduction/reduction_span_pack.cpp

```diff
@@ -0,0 +1,226 @@
+      } else {
+        Passed &= (Output2[I] == ExpectedRemainder);
+      }
+    }
+  }
+
+  free(Output2, Q);
+  free(Output1, Q);
+  NumErrors += (Passed) ? 0 : 1;
+}
+
+struct CustomType {
+  int x;
+  bool operator==(const CustomType &o) const { return (x == o.x); }
+};
+
+struct CustomBinaryOperation {
+  CustomType operator()(const CustomType &lhs, const CustomType &rhs) const {
+    return CustomType{lhs.x + rhs.x};
+  }
+};
+
+template <size_t N, typename T, typename BinaryOperation, typename Range,
+          submission_mode SubmissionMode>
+void test(queue Q, Range Rng, T Identity, T Value) {
+  test1<N, T, BinaryOperation, Range, SubmissionMode>(Q, Rng, Identity, Value);
+  test2<N, T, BinaryOperation, Range, SubmissionMode>(Q, Rng, Identity, Value);
+}
+
+int main() {
+  queue Q;
+
+  // Tests for small spans that can be privatized efficiently
+  // Each combination tests a different sycl::reduction implementation
+  // TODO: Enable range<> tests once parallel_for accepts pack
+  /*test<16, int, std::plus<int>, sycl::range<1>, submission_mode::handler>(Q,
+  24, 0, 1); test<16, float, std::plus<float>, sycl::range<1>,
+  submission_mode::handler>(Q, 24, 0, 1); test<16, int, std::multiplies<int>,
+  sycl::range<1>, submission_mode::handler>(Q, 24, 1, 2); test<16, CustomType,
+  CustomBinaryOperation, sycl::range<1>, submission_mode::handler>(Q, 24,
+  CustomType{0}, CustomType{1});
+  test<16, int, std::plus<int>, sycl::range<1>, submission_mode::queue>(Q, 24,
+  0, 1); test<16, float, std::plus<float>, sycl::range<1>,
+  submission_mode::queue>(Q, 24, 0, 1); test<16, int, std::multiplies<int>,
+  sycl::range<1>, submission_mode::queue>(Q, 24, 1, 2); test<16, CustomType,
+  CustomBinaryOperation, sycl::range<1>, submission_mode::queue>(Q, 24,
+  CustomType{0}, CustomType{1});*/
+
+  test<16, int, std::plus<int>, sycl::nd_range<1>, submission_mode::handler>(
+      Q, {24, 8}, 0, 1);
+  test<16, float, std::plus<float>, sycl::nd_range<1>,
+       submission_mode::handler>(Q, {24, 8}, 0, 1);
+  test<16, int, std::multiplies<int>, sycl::nd_range<1>,
+       submission_mode::handler>(Q, {24, 8}, 1, 2);
+  test<16, int, std::bit_or<int>, sycl::nd_range<1>, submission_mode::handler>(
+      Q, {24, 8}, 0, 1);
+  test<16, CustomType, CustomBinaryOperation, sycl::nd_range<1>,
+       submission_mode::handler>(Q, {24, 8}, CustomType{0}, CustomType{1});
+  test<16, int, std::plus<int>, sycl::nd_range<1>, submission_mode::queue>(
+      Q, {24, 8}, 0, 1);
+  test<16, float, std::plus<float>, sycl::nd_range<1>, submission_mode::queue>(
+      Q, {24, 8}, 0, 1);
+  test<16, int, std::multiplies<int>, sycl::nd_range<1>,
+       submission_mode::queue>(Q, {24, 8}, 1, 2);
+  test<16, int, std::bit_or<int>, sycl::nd_range<1>, submission_mode::queue>(
+      Q, {24, 8}, 0, 1);
+  test<16, CustomType, CustomBinaryOperation, sycl::nd_range<1>,
+       submission_mode::queue>(Q, {24, 8}, CustomType{0}, CustomType{1});
```

Comment:
Minor comment:
It seems like the lines 204-213 and 214-223 have only 1 difference - the submission_mode.
If so, it could be easier/shorter to add 2 lines near the line 182, i.e.
```
template <size_t N, typename T, typename BinaryOperation, typename Range>
void test(queue Q, Range Rng, T Identity, T Value) {
  test1<N, T, BinaryOperation, Range, submission_mode::handler>(Q, Rng, Identity, Value);
  test2<N, T, BinaryOperation, Range, , submission_mode::handler>(Q, Rng, Identity, Value);
  test1<N, T, BinaryOperation, Range, , submission_mode::queue>(Q, Rng, Identity, Value);
  test2<N, T, BinaryOperation, Range, submission_mode::queue>(Q, Rng, Identity, Value);
}
```